src/Lecture2.hs

```diff
@@ -1,170 +1,292 @@
+--     4. Add polymorphism.
+--     5. Make all invalid states unrepresentable. Think, how you can
+--        change your types to prevent green dragons from having any
+--        treasure besides gold (if you already haven't done this).
 
 -- some help in the beginning ;)
-data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
-    }
 
-dragonFight = error "TODO"
+data Knight = Knight
+  { knightHealth :: Health,
+    knightAttack :: Attack,
+    knightEndurance :: Endurance
+  }
+
+data Dragon a
+  = RedDragon DragonStats (Chest a)
+  | BlackDragon DragonStats (Chest a)
+  | GreenDragon DragonStats Gold
+
+data DragonStats = DragonStats
+  { dragonStatsHealth :: Health,
+    dragonStatsAttack :: Attack
+  }
+
+dragonStats :: Dragon a -> DragonStats
+dragonStats (RedDragon stats _) = stats
+dragonStats (BlackDragon stats _) = stats
+dragonStats (GreenDragon stats _) = stats
+
+dragonHealth :: Dragon a -> Health
+dragonHealth = dragonStatsHealth . dragonStats
+
+dragonAttack :: Dragon a -> Attack
+dragonAttack = dragonStatsAttack . dragonStats
+
+setDragonStats :: DragonStats -> Dragon a -> Dragon a
+setDragonStats s d = case d of
+  RedDragon _ loot -> RedDragon s loot
+  BlackDragon _ loot -> BlackDragon s loot
+  GreenDragon _ loot -> GreenDragon s loot
+
+data Chest a = Chest
+  { chestGold :: Gold,
+    chestTreasure :: Treasure a
+  }
+
+data Treasure a = NoTreasure | SomeTreasure a
+
+data Reward a = Reward
+  { rewardExp :: Experience,
+    rewardGold :: Gold,
+    rewardTreasure :: Treasure a
+  }
+
+data FightResult a
+  = KnightWin (Reward a)
+  | KnightLose
+  | KnightFlee
+
+-- Newtypes to avoid ambiguities in the calculateDamage phase
+newtype Attack = Attack {unAttack :: Int}
+
+-- NOTE (Student) Deriving typeclasses makes things easier for calculations
+-- but we will use records members for now instead of
+-- deriving (Num, Eq, Ord)
+newtype Health = Health {unHealth :: Int}
+
+type Endurance = Int
+
+-- NOTE (Student) Ideally, we should represent the fact that
+-- gold is a positive int. I found the following answer from
+-- Gabriella Gonzalez https://stackoverflow.com/a/11910221
+-- However, it doesn't seem possible to prevent creating a Positive Int
+-- with a wrapped negative int if we are in the same module.
+type Gold = Int
+
+type Experience = Int
+
+calculateRewardExp :: Dragon a -> Experience
+calculateRewardExp d = case d of
+  RedDragon {} -> 100
+  BlackDragon {} -> 150
+  GreenDragon {} -> 250
+
+calculateReward :: Dragon a -> Reward a
+calculateReward d = case d of
+  GreenDragon _ gold -> Reward {rewardExp = calculateRewardExp d, rewardGold = gold, rewardTreasure = NoTreasure}
+  BlackDragon _ chest -> Reward {rewardExp = calculateRewardExp d, rewardGold = chestGold chest, rewardTreasure = chestTreasure chest}
+  RedDragon _ chest -> Reward {rewardExp = calculateRewardExp d, rewardGold = chestGold chest, rewardTreasure = chestTreasure chest}
```

Comment:
To reduce some boilerplate in this function, you can write a separate function to get only the treasure reward from a dragon:

```haskell
calculateRewardTreasure :: Dragon a -> (Gold, Treasure a)
```

After that, you'll be able to write this function shorter:

```haskell
calculateReward :: Dragon a -> Reward a
calculateReward d = 
    let (gold, treasure) = calculateRewardTreasure d in
    Reward
        { rewardExp = calculateRewardExp d
        , rewardGold = gold,
        , rewardTreasure = treasure
        }
```

Since you're prefixing record fields with the type name, you can even benefit from a handy Haskell extension [RecordWildCards](https://kodimensional.dev/recordwildcards) and write the above function in a slightly shorter way:

```haskell
calculateReward :: Dragon a -> Reward a
calculateReward d = 
    let (rewardGold, rewardTeasure) = calculateRewardTreasure d in
    Reward { rewardExp = calculateRewardExp d, .. }
```

Reply:
That is a lot cleaner indeed ! I wrote the function to get the experience but haven't thought about doing the same for the treasure...
I'll keep the RecordWildCards extension in mind, it seems really useful when used correctly. Great article by the way !